### PR TITLE
PLANNER-2541 Simplify ProblemFactChange interface

### PR DESCRIPTION
**Pull request by rsynek**
### JIRA
https://issues.redhat.com/browse/PLANNER-2541

### Referenced pull requests
https://github.com/kiegroup/optaplanner/pull/1703
https://github.com/kiegroup/optaplanner-quickstarts/pull/259
https://github.com/kiegroup/optaweb-vehicle-routing/pull/635

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* for a <b>pull request build</b> please add comment: <b>Jenkins retest this</b>
* for a <b>specific pull request build</b> please add comment: <b>Jenkins (re)run [optaweb-vehicle-routing] tests</b>
* for a <b>full downstream build</b> 
  * for <b>jenkins</b> job: please add comment: <b>Jenkins run fdb</b>
  * for <b>github actions</b> job: add the label `run_fdb`
* for a <b>compile downstream build</b> please add comment: <b>Jenkins run cdb</b>
* for a <b>full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>
* for an <b>upstream build</b> please add comment: <b>Jenkins run upstream</b>
* for a <b>Quarkus LTS check</b> please add comment: <b>Jenkins run LTS</b>
* for a <b>specific Quarkus LTS check</b> please add comment: <b>Jenkins (re)run [optaweb-vehicle-routing] LTS</b>

</details>

### CODE REVIEW

File: optaweb-vehicle-routing-backend/src/test/java/org/optaweb/vehiclerouting/plugin/planner/change/ChangeVehicleCapacityTest.java [hunk 1]
```diff
@@ -17,51 +17,33 @@
...
 
         assertThat(workingVehicle.getCapacity()).isEqualTo(newCapacity);
-
-        verify(scoreDirector).beforeProblemPropertyChanged(workingVehicle);
-        verify(scoreDirector).afterProblemPropertyChanged(workingVehicle);
-        verify(scoreDirector).triggerVariableListeners();
-    }
-
-    @Test
-    void fail_fast_if_working_object_is_null() {
```
Comment (rsynek) [inline optaweb-vehicle-routing-backend/src/test/java/org/optaweb/vehiclerouting/plugin/planner/change/ChangeVehicleCapacityTest.java:61]: The `ProblemChangeDirector` fails fast if the object it's adding, removing or changing is null. Thus, we no longer need this test as the `ProblemChange` implementation doesn't have to care about it.

File: optaweb-vehicle-routing-backend/src/test/java/org/optaweb/vehiclerouting/plugin/planner/change/AddVehicleTest.java [hunk 2]
```diff
@@ -17,38 +17,28 @@
...
 class AddVehicleTest {
 
-    @Mock
-    private ScoreDirector<VehicleRoutingSolution> scoreDirector;
-
     @Test
     void add_vehicle_should_add_vehicle() {
         VehicleRoutingSolution solution = SolutionFactory.emptySolution();
-        when(scoreDirector.getWorkingSolution()).thenReturn(solution);
+        MockProblemChangeDirector mockProblemChangeDirector = spy(new MockProblemChangeDirector());
```
Comment (rsynek) [inline optaweb-vehicle-routing-backend/src/test/java/org/optaweb/vehiclerouting/plugin/planner/change/AddVehicleTest.java:34]: A new class coming from the `optaplanner-test` module; its role is to provide look-ups and call the lambdas defined in the `ProblemChange` implementation.

File: optaweb-vehicle-routing-backend/src/test/java/org/optaweb/vehiclerouting/plugin/planner/SolverManagerTest.java [hunk 3]
```diff
@@ -21,9 +21,7 @@
 import static org.assertj.core.api.Assertions.assertThatIllegalStateException;
 import static org.mockito.AdditionalAnswers.answer;
 import static org.mockito.ArgumentMatchers.any;
-import static org.mockito.Mockito.never;
-import static org.mockito.Mockito.verify;
-import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.*;
```
Comment (yurloc) [inline optaweb-vehicle-routing-backend/src/test/java/org/optaweb/vehiclerouting/plugin/planner/SolverManagerTest.java:24]: Please avoid star imports.

File: optaweb-vehicle-routing-backend/src/main/java/org/optaweb/vehiclerouting/plugin/planner/change/RemoveVehicle.java [hunk 4]
```diff
@@ -33,39 +33,34 @@ public RemoveVehicle(PlanningVehicle removedVehicle) {
...
-    public void doChange(ScoreDirector<VehicleRoutingSolution> scoreDirector) {
-        VehicleRoutingSolution workingSolution = scoreDirector.getWorkingSolution();
-
+    public void doChange(VehicleRoutingSolution workingSolution, ProblemChangeDirector problemChangeDirector) {
         // Look up a working copy of the vehicle
-        PlanningVehicle workingVehicle = scoreDirector.lookUpWorkingObject(removedVehicle);
+        PlanningVehicle workingVehicle = problemChangeDirector.lookUpWorkingObject(removedVehicle);
         if (workingVehicle == null) {
             throw new IllegalStateException("Can't look up a working copy of " + removedVehicle);
         }
```
Comment (yurloc) [inline optaweb-vehicle-routing-backend/src/main/java/org/optaweb/vehiclerouting/plugin/planner/change/RemoveVehicle.java:41]: @rsynek This is obsolete as in the capacity change, right?

Comment (yurloc) [reply optaweb-vehicle-routing-backend/src/main/java/org/optaweb/vehiclerouting/plugin/planner/change/RemoveVehicle.java:41]: The check was probably redundant from the beginning. Now's a good time to get rid of it.

**Review by yurloc — APPROVED** [review]

**Pull request merged**